### PR TITLE
Tpetra: Remove use of deprecated Kokkos code

### PR DESCRIPTION
**Pull request by csiefer2**
Issue: #11510
Tests: Pass
Stakeholder feedback: Discussion on Kokkos slack, See also #11510

@dalg24

### CODE REVIEW

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp [hunk 1]
```diff
@@ -366,6 +366,7 @@ namespace Impl {
     KOKKOS_INLINE_FUNCTION void
     operator () (const typename Kokkos::TeamPolicy<typename device_type::execution_space>::member_type & member) const
     {
+      constexpr bool runOnHost = std::is_same<typename Kokkos::TeamPolicy<typename device_type::execution_space>::member_type,Kokkos::DefaultHostExecutionSpace>::value;
```
Comment (dalg24) [inline packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp:369]:
```suggestion
      constexpr bool runOnHost = !std::is_same_v<typename Kokkos::TeamPolicy<typename device_type::execution_space>::member_type, Kokkos::DefaultExecutionSpace> || std::is_same_v<Kokkos::DefaultExecutionSpace, Kokkos::DefaultHostExecutionSpace>;
```
```suggestion
      constexpr bool runOnHost = std::is_same<typename Kokkos::TeamPolicy<typename device_type::execution_space>::member_type,Kokkos::DefaultHostExecutionSpace>::value;
```

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp [hunk 2]
```diff
@@ -424,16 +425,17 @@ namespace Impl {
...
-                // host space team size is always 1
-                Y_cur(k0) += alpha_*val;
-#else
-                // cuda space team size can be larger than 1
-                // atomic is not allowed for sacado type;
-                // thus this needs to be specialized or 
-                // sacado atomic should be supported.
-                Kokkos::atomic_add(&Y_cur(k0), alpha_*val);
-#endif
+                if constexpr(runOnHost) {
```
Comment (ndellingwood) [inline packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp:428]: Woo-hoo for c++17!

**Review by ndellingwood — APPROVED** [review]

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp [hunk 3]
```diff
@@ -366,6 +366,7 @@ namespace Impl {
     KOKKOS_INLINE_FUNCTION void
     operator () (const typename Kokkos::TeamPolicy<typename device_type::execution_space>::member_type & member) const
     {
+      constexpr bool runOnHost = !std::is_same_v<typename Kokkos::TeamPolicy<typename device_type::execution_space>::member_type, Kokkos::DefaultExecutionSpace> || std::is_same_v<Kokkos::DefaultExecutionSpace, Kokkos::DefaultHostExecutionSpace>;
```
Comment (crtrott) [inline packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp:369]: this doesn't make sense??

**Review by crtrott — CHANGES_REQUESTED** [review]

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp (hunk 3 above)
Comment (dalg24) [inline packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp:369]:
```suggestion
      constexpr bool runOnHost = !std::is_same_v<typename device_type::execution_space, Kokkos::DefaultExecutionSpace> || std::is_same_v<Kokkos::DefaultExecutionSpace, Kokkos::DefaultHostExecutionSpace>;
```

Comment (crtrott) [reply packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp:369]: is this meant to be:
```c++
constexpr bool runOnHost = !std::is_same_v<typename device_type::execution_space, Kokkos::DefaultExecutionSpace> || std::is_same_v<Kokkos::DefaultExecutionSpace, Kokkos::DefaultHostExecutionSpace>
```

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp (hunk 2 above)
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp:428]: Indeed!

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp (hunk 3 above)
Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp]: I didn't know github can do code suggestions like how @dalg24 has been doing.  Very cool!

**Review by crtrott — APPROVED** [review]

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp [hunk 4]
```diff
@@ -366,6 +371,7 @@ namespace Impl {
     KOKKOS_INLINE_FUNCTION void
     operator () (const typename Kokkos::TeamPolicy<typename device_type::execution_space>::member_type & member) const
     {
+      //      constexpr bool runOnHost = !std::is_same_v<typename device_type::execution_space, Kokkos::DefaultExecutionSpace> || std::is_same_v<Kokkos::DefaultExecutionSpace, Kokkos::DefaultHostExecutionSpace>;
```
Comment (lucbv) [inline packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp:374]: This should be removed

Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_BlockCrsMatrix_def.hpp]: Done

**Review by lucbv — APPROVED** [review]
Looks good to me, thanks

**Pull request merged**